File: opencga-client/src/main/python/pyOpenCGA/pyopencga/rest_clients/alignment_client.py
from pyopencga.rest_clients._parent_rest_clients import _ParentRestClient
from pyopencga.commons import deprecated


class Alignment(_ParentRestClient):
    """
    This class contains methods for the AnalysisAlignment webservices
    """

    def __init__(self, configuration, token=None, login_handler=None, *args, **kwargs):
        _category = 'analysis/alignment'
        super(Alignment, self).__init__(configuration, _category, token, login_handler, *args, **kwargs)

    def index(self, file, **options):
        """
        Index alignment files
        URL: /{apiVersion}/analysis/alignment/index

        :param file: Comma separated list of file ids (files or directories)
        """

        options['file'] = file

        return self._post('index', **options)

    @deprecated
    def stats(self, file, **options):
        return self.stats_info(file, **options)

    def run_stats(self, file, **options):
        """
        Compute stats for a given alignment file
        URL: /{apiVersion}/analysis/alignment/stats/run

        :param file: file id or name in Catalog
        """

        options['file'] = file

        return self._post('stats/run', **options)

    def stats_info(self, file, **options):
        """
        Show the stats for a given alignment file
        URL: /{apiVersion}/analysis/alignment/stats/info

        :param file: file id or name in Catalog
        """

        options['file'] = file

        return self._get('stats/info', **options)

    def query_stats(self, **options):
        """
        Fetch alignment files according to their stats
        URL: /{apiVersion}/analysis/alignment/stats/query
        """

        return self._get('stats/query', **options)

    def run_bwa(self, data, **options):
        """
        BWA is a software package for mapping low-divergent sequences against a
        large reference genome.
        URL: /{apiVersion}/analysis/alignment/bwa/run
        """

        return self._post('bwa/run', data=data, **options)

    def run_samtools(self, data, **options):
        """
        Samtools is a program for interacting with high-throughput sequencing
        data in SAM, BAM and CRAM formats.
        URL: /{apiVersion}/analysis/alignment/samtools/run
        """

        return self._post('samtools/run', data=data, **options)

    def run_deeptools(self, data, **options):
        """
        Deeptools is a suite of python tools particularly developed for the
        efficient analysis of high-throughput sequencing data, such as ChIP-seq,
        RNA-seq or MNase-seq.
        URL: /{apiVersion}/analysis/alignment/deeptools/run
        """

        return self._post('deeptools/run', data=data, **options)

    def run_fastqc(self, data, **options):
        """
        A quality control tool for high throughput sequence data
        URL: /{apiVersion}/analysis/alignment/fastqc/run
        """

        return self._post('fastqc/run', data=data, **options)

    def query(self, file, **options):
        """
        Search over indexed alignments
        URL: /{apiVersion}/analysis/alignment/query

        :param file: file id or name in Catalog
        :param region: comma-separated list of regions 'chr:start-end'
        """

        options['file'] = file

        return self._get('query', **options)

    def run_coverage(self, file, **options):
        """
        Compute coverage for a list of alignment files
        URL: /{apiVersion}/analysis/alignment/coverage/run

        :param file: file id or name in Catalog
        """

        options['file'] = file

        return self._post('coverage/run', **options)

    def query_coverage(self, file, **options):
        """
        Query the coverage of an alignment file for regions or genes
        URL: /{apiVersion}/analysis/alignment/coverage/query

        :param file: file id or name in Catalog
        """

        options['file'] = file

        return self._get('coverage/query', **options)

    def coverage_ratio(self, file1, file2, **options):
        """
        Compute coverage ratio from file #1 vs file #2, (e.g. somatic vs germline)
        URL: /{apiVersion}/analysis/alignment/coverage/ratio

        :param file1: file1 id or name in Catalog
        :param file2: file2 id or name in Catalog
        """

        options['file1'] = file1
        options['file2'] = file2

        return self._get('coverage/ratio', **options)

File: opencga-client/src/main/java/org/opencb/opencga/client/rest/AlignmentClient.java
/*
* Copyright 2015-2020 OpenCB
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

package org.opencb.opencga.client.rest;

import org.ga4gh.models.ReadAlignment;
import org.opencb.biodata.models.alignment.RegionCoverage;
import org.opencb.commons.datastore.core.ObjectMap;
import org.opencb.opencga.client.config.ClientConfiguration;
import org.opencb.opencga.client.exceptions.ClientException;
import org.opencb.opencga.core.models.file.File;
import org.opencb.opencga.core.models.job.Job;
import org.opencb.opencga.core.response.RestResponse;


/**
 * This class contains methods for the Alignment webservices.
 *    Client version: 2.0.0
 *    PATH: analysis/alignment
 */
public class AlignmentClient extends AbstractParentClient {

    public AlignmentClient(String token, ClientConfiguration configuration) {
        super(token, configuration);
    }

    /**
     * Compute coverage for a list of alignment files.
     * @param file File ID.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runCoverage(String file, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file", file);
        return execute("analysis/alignment", null, "coverage", null, "run", params, POST, Job.class);
    }

    /**
     * Query the coverage of an alignment file for regions or genes.
     * @param file File ID.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<RegionCoverage> queryCoverage(String file, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file", file);
        return execute("analysis/alignment", null, "coverage", null, "query", params, GET, RegionCoverage.class);
    }

    /**
     * Compute coverage ratio from file #1 vs file #2, (e.g. somatic vs germline).
     * @param file1 Input file #1 (e.g. somatic file).
     * @param file2 Input file #2 (e.g. germline file).
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<RegionCoverage> ratioCoverage(String file1, String file2, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file1", file1);
        params.putIfNotNull("file2", file2);
        return execute("analysis/alignment", null, "coverage", null, "ratio", params, GET, RegionCoverage.class);
    }

    /**
     * Compute stats for a given alignment file.
     * @param file File ID.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runStats(String file, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file", file);
        return execute("analysis/alignment", null, "stats", null, "run", params, POST, Job.class);
    }

    /**
     * Show the stats for a given alignment file.
     * @param file File ID.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<String> infoStats(String file, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file", file);
        return execute("analysis/alignment", null, "stats", null, "info", params, GET, String.class);
    }

    /**
     * Fetch alignment files according to their stats.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<File> queryStats(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/alignment", null, "stats", null, "query", params, GET, File.class);
    }

    /**
     * BWA is a software package for mapping low-divergent sequences against a large reference genome.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runBwa(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/alignment", null, "bwa", null, "run", params, POST, Job.class);
    }

    /**
     * Samtools is a program for interacting with high-throughput sequencing data in SAM, BAM and CRAM formats.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runSamtools(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/alignment", null, "samtools", null, "run", params, POST, Job.class);
    }

    /**
     * Deeptools is a suite of python tools particularly developed for the efficient analysis of high-throughput sequencing data, such as
     *     ChIP-seq, RNA-seq or MNase-seq.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runDeeptools(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/alignment", null, "deeptools", null, "run", params, POST, Job.class);
    }

    /**
     * A quality control tool for high throughput sequence data.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runFastqc(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/alignment", null, "fastqc", null, "run", params, POST, Job.class);
    }

    /**
     * Index alignment file.
     * @param file File ID.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> index(String file, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file", file);
        return execute("analysis/alignment", null, null, null, "index", params, POST, Job.class);
    }

    /**
     * Search over indexed alignments.
     * @param file File ID.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ReadAlignment> query(String file, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.putIfNotNull("file", file);
        return execute("analysis/alignment", null, null, null, "query", params, GET, ReadAlignment.class);
    }
}
